Reproducible Failure in org.elasticsearch.search.aggregations.metrics.InternalTopHitsTests.testReduceRandom

This reproduces 100% of the time for me locally on current master:

```
Suite: Test class org.elasticsearch.search.aggregations.metrics.InternalTopHitsTests
  1> [2020-11-17T21:21:56,451][INFO ][o.e.s.a.m.InternalTopHitsTests] [testReduceRandom] before test
  1> [2020-11-17T21:21:56,586][INFO ][o.e.s.a.m.InternalTopHitsTests] [testReduceRandom] after test
  2> REPRODUCE WITH: ./gradlew ':server:test' --tests "org.elasticsearch.search.aggregations.metrics.InternalTopHitsTests.testReduceRandom" -Dtests.seed=3F5182C50EE4E338 -Dtests.security.manager=true -Dtests.locale=sk-SK -Dtests.timezone=Europe/Andorra -Druntime.java=15
  2> java.lang.AssertionError: Didn't match expected value:
                              hits: 
                                hits: 
                                     0: 
                                     _id: same [9]
                                  _score: same [0.99827045]
                                     1: 
                                     _id: same [10]
                                  _score: same [0.996799]
                                     2: 
                                     _id: same [0]
                                  _score: same [0.9966625]
                                     3: 
                                     _id: same [18]
                                  _score: same [0.99489796]
                                     4: 
                                     _id: same [11]
                                  _score: same [0.9931389]
                                     5: 
                                     _id: same [7]
                                  _score: same [0.99165297]
                                     6: 
                                     _id: same [3]
                                  _score: same [0.9915664]
                                     7: 
                                     _id: same [12]
                                  _score: same [0.9913156]
                                     8: 
                                     _id: same [0]
                                  _score: same [0.9905108]
                                     9: 
                                     _id: same [3]
                                  _score: same [0.9893769]
                                    10: 
                                     _id: same [1]
                                  _score: same [0.98848015]
                                    11: 
                                     _id: expected String [1] but was String [13]
                                  _score: expected Double [0.9830506] but was Double [0.98169035]
                           max_score: same [0.99827045]
                               total: 
                              relation: same [eq]
                                 value: same [24404932]
        at __randomizedtesting.SeedInfo.seed([3F5182C50EE4E338:FF6C2ED428B97A6E]:0)
        at org.elasticsearch.search.aggregations.metrics.InternalTopHitsTests.assertEqualsWithErrorMessageFromXContent(InternalTopHitsTests.java:298)
        at org.elasticsearch.search.aggregations.metrics.InternalTopHitsTests.assertReduced(InternalTopHitsTests.java:269)
        at org.elasticsearch.search.aggregations.metrics.InternalTopHitsTests.assertReduced(InternalTopHitsTests.java:67)
  2> NOTE: leaving temporary files on disk at: /home/brownbear/src/elasticsearch/server/build/testrun/test/temp/org.elasticsearch.search.aggregations.metrics.InternalTopHitsTests_3F5182C50EE4E338-002
  2> NOTE: test params are: codec=Lucene87, sim=Asserting(RandomSimilarity(queryNorm=false): {}), locale=sk-SK, timezone=Europe/Andorra
  2> NOTE: Linux 5.4.0-54-generic amd64/AdoptOpenJDK 15.0.1 (64-bit)/cpus=16,threads=1,free=478926608,total=536870912
  2> NOTE: All tests run in this JVM: [InternalTopHitsTests]


```

--- a/server/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalTopHitsTests.java
+++ b/server/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalTopHitsTests.java
@@ -68,10 +68,10 @@ public class InternalTopHitsTests extends InternalAggregationTestCase<InternalTo
     @Override
     protected InternalTopHits createTestInstance(String name, Map<String, Object> metadata) {
         if (randomBoolean()) {
-            return createTestInstanceSortedByFields(name, metadata, ESTestCase::randomFloat,
+            return createTestInstanceSortedByFields(name, between(1, 40), metadata, ESTestCase::randomFloat,
                 randomSortFields(), InternalTopHitsTests::randomOfType);
         }
-        return createTestInstanceSortedScore(name, metadata, ESTestCase::randomFloat);
+        return createTestInstanceSortedScore(name, between(1, 40), metadata, ESTestCase::randomFloat);
     }
 
     @Override
@@ -86,6 +86,7 @@ public class InternalTopHitsTests extends InternalAggregationTestCase<InternalTo
             usedScores.add(score);
             return score;
         };
+        int requestedSize = between(1, 40);
         Supplier<InternalTopHits> supplier;
         if (randomBoolean()) {
             SortField[] sortFields = randomSortFields();
@@ -95,16 +96,16 @@ public class InternalTopHitsTests extends InternalAggregationTestCase<InternalTo
                 usedSortFieldValues.add(value);
                 return value;
             };
-            supplier = () -> createTestInstanceSortedByFields(name, null, scoreSupplier, sortFields, sortFieldValueSuppier);
+            supplier = () -> createTestInstanceSortedByFields(name, requestedSize, null, scoreSupplier, sortFields, sortFieldValueSuppier);
         } else {
-            supplier = () -> createTestInstanceSortedScore(name, null, scoreSupplier);
+            supplier = () -> createTestInstanceSortedScore(name, requestedSize, null, scoreSupplier);
         }
         return Stream.generate(supplier).limit(size).collect(toList());
     }
 
-    private InternalTopHits createTestInstanceSortedByFields(String name, Map<String, Object> metadata,
+    private InternalTopHits createTestInstanceSortedByFields(String name, int requestedSize, Map<String, Object> metadata,
             Supplier<Float> scoreSupplier, SortField[] sortFields, Function<SortField.Type, Object> sortFieldValueSupplier) {
-        return createTestInstance(name, metadata, scoreSupplier,
+        return createTestInstance(name, metadata, scoreSupplier, requestedSize,
             (docId, score) -> {
                 Object[] fields = new Object[sortFields.length];
                 for (int f = 0; f < sortFields.length; f++) {
@@ -117,15 +118,15 @@ public class InternalTopHitsTests extends InternalAggregationTestCase<InternalTo
             sortFieldsComparator(sortFields));
     }
 
-    private InternalTopHits createTestInstanceSortedScore(String name, Map<String, Object> metadata, Supplier<Float> scoreSupplier) {
-        return createTestInstance(name, metadata, scoreSupplier, ScoreDoc::new, TopDocs::new, scoreComparator());
+    private InternalTopHits createTestInstanceSortedScore(String name, int requestedSize, Map<String, Object> metadata, Supplier<Float> scoreSupplier) {
+        return createTestInstance(name, metadata, scoreSupplier, requestedSize, ScoreDoc::new, TopDocs::new, scoreComparator());
     }
 
-    private InternalTopHits createTestInstance(String name, Map<String, Object> metadata, Supplier<Float> scoreSupplier, 
+    private InternalTopHits createTestInstance(String name, Map<String, Object> metadata, Supplier<Float> scoreSupplier,
+            int requestedSize,
             BiFunction<Integer, Float, ScoreDoc> docBuilder,
             BiFunction<TotalHits, ScoreDoc[], TopDocs> topDocsBuilder, Comparator<ScoreDoc> comparator) {
         int from = 0;
-        int requestedSize = between(1, 40);
         int actualSize = between(0, requestedSize);
 
         float maxScore = Float.NEGATIVE_INFINITY;

--- a/server/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalTopHitsTests.java
+++ b/server/src/test/java/org/elasticsearch/search/aggregations/metrics/InternalTopHitsTests.java
@@ -118,7 +118,12 @@ public class InternalTopHitsTests extends InternalAggregationTestCase<InternalTo
             sortFieldsComparator(sortFields));
     }
 
-    private InternalTopHits createTestInstanceSortedScore(String name, int requestedSize, Map<String, Object> metadata, Supplier<Float> scoreSupplier) {
+    private InternalTopHits createTestInstanceSortedScore(
+        String name,
+        int requestedSize,
+        Map<String, Object> metadata,
+        Supplier<Float> scoreSupplier
+    ) {
         return createTestInstance(name, metadata, scoreSupplier, requestedSize, ScoreDoc::new, TopDocs::new, scoreComparator());
     }
 

--- a/test/framework/src/main/java/org/elasticsearch/test/InternalAggregationTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/test/InternalAggregationTestCase.java
@@ -333,7 +333,6 @@ public abstract class InternalAggregationTestCase<T extends InternalAggregation>
     }
 
     @Override
-
     protected final Class<T> categoryClass() {
         return (Class<T>) InternalAggregation.class;
     }
@@ -353,7 +352,6 @@ public abstract class InternalAggregationTestCase<T extends InternalAggregation>
         return inputs;
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/65163")
     public void testReduceRandom() throws IOException {
         String name = randomAlphaOfLength(5);
         int size = between(1, 200);

--- a/x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetricsTests.java
+++ b/x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetricsTests.java
@@ -6,6 +6,8 @@
 
 package org.elasticsearch.xpack.analytics.topmetrics;
 
+import com.carrotsearch.randomizedtesting.annotations.Repeat;
+
 import org.elasticsearch.client.analytics.ParsedTopMetrics;
 import org.elasticsearch.common.ParseField;
 import org.elasticsearch.common.Strings;
@@ -328,6 +330,27 @@ public class InternalTopMetricsTests extends InternalAggregationTestCase<Interna
                 }
             }
         }
+    }
+
+    @Override
+    protected List<InternalTopMetrics> randomResultsToReduce(String name, int count) {
+        InternalTopMetrics prototype = createTestInstance();
+        return randomList(
+            count,
+            count,
+            () -> new InternalTopMetrics(
+                prototype.getName(),
+                prototype.getSortOrder(),
+                prototype.getMetricNames(),
+                prototype.getSize(),
+                randomTopMetrics(
+                    InternalAggregationTestCase::randomNumericDocValueFormat,
+                    between(0, prototype.getSize()),
+                    prototype.getMetricNames().size()
+                ),
+                prototype.getMetadata()
+            )
+        );
     }
 
     @Override

--- a/x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetricsTests.java
+++ b/x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetricsTests.java
@@ -6,8 +6,6 @@
 
 package org.elasticsearch.xpack.analytics.topmetrics;
 
-import com.carrotsearch.randomizedtesting.annotations.Repeat;
-
 import org.elasticsearch.client.analytics.ParsedTopMetrics;
 import org.elasticsearch.common.ParseField;
 import org.elasticsearch.common.Strings;

